micrometer-core/src/main/java/io/micrometer/core/instrument/HighCardinalityTagsDetector.java

```diff
@@ -0,0 +1,179 @@
+        this.scheduledExecutorService = Executors
+                .newSingleThreadScheduledExecutor(new NamedThreadFactory("high-cardinality-tags-detector"));
+    }
+
+    /**
+     * Starts a scheduled job that checks if you have high cardinality tags.
+     */
+    public void start() {
+        LOGGER.info(String.format("Starting %s with threshold: %d and delay: %s", this.getClass().getSimpleName(),
+                this.threshold, this.delay));
+        this.scheduledExecutorService.scheduleWithFixedDelay(this::detectHighCardinalityTags, 0, this.delay.toMillis(),
+                TimeUnit.MILLISECONDS);
+    }
+
+    /**
+     * Shuts down the scheduled job that checks if you have high cardinality tags.
+     */
+    public void shutdown() {
+        LOGGER.info("Stopping " + this.getClass().getSimpleName());
+        this.scheduledExecutorService.shutdown();
+    }
+
+    @Override
+    public void close() {
+        this.shutdown();
+    }
+
+    private void detectHighCardinalityTags() {
+        try {
+            findFirst().ifPresent(this.meterNameConsumer);
+        }
+        catch (Exception exception) {
+            LOGGER.warn("Something went wrong during high cardinality tag detection", exception);
+        }
+    }
+
+    /**
+     * Finds the name of the first Meter that potentially has high cardinality tags.
+     * @return the name of the first Meter that potentially has high cardinality tags, an
+     * empty Optional if none found.
+     */
+    public Optional<String> findFirst() {
+        Map<String, Long> meterNameFrequencies = new HashMap<>();
+        for (Meter meter : this.registry.getMeters()) {
+            String name = meter.getId().getName();
+            if (!meterNameFrequencies.containsKey(name)) {
+                meterNameFrequencies.put(name, 1L);
+            }
+            else {
+                Long frequency = meterNameFrequencies.get(name);
+                if (frequency < this.threshold) {
+                    meterNameFrequencies.put(name, frequency + 1);
+                }
+                else {
+                    return Optional.of(name);
+                }
+            }
+        }
+
+        return Optional.empty();
+    }
+
+    private void logWarning(String name) {
+        WARN_THEN_DEBUG_LOGGER.log(String.format("It seems %s has high cardinality tags (threshold: %d meters).\n"
+                + "Check your configuration for the instrumentation of %s to find and fix the cause of the high cardinality (see: https://micrometer.io/docs/concepts#_tag_values).\n"
+                + "If the cardinality is expected and acceptable, raise the threshold for this %s.", name,
+                this.threshold, name, this.getClass().getSimpleName()));
+    }
+
+    private static long calculateThreshold() {
+        // half of the heap in MiB
+        long allowance = Runtime.getRuntime().maxMemory() / 1024 / 1024 / 2;
```

Comment:
I wonder if half the max heap isn't too much for one meter name to be allowed to take before warning, considering there will be other meters too. This is a bit of a bike shed issue on what the right default is, though. I would think keeping total metrics memory overhead to 10% or less of the heap would be ideal. However, this doesn't count against all metrics registered, just metrics with the same name in one registry. If there is more than one registry, 50% * 2 is already and OOM situation.

Reply:
I think lowering the threshold is a good idea. I think it also worth calling out in the javadoc that this component will not detect if Micrometer is using a lot of memory, it only tries to detect if the app potentially has high cardinality tags. E.g.: it won't detect if one has no tags but a timestamp is concatenated to the name.

We can implement a `HighMemoryUsageDetector` or make this more generic (see: https://github.com/micrometer-metrics/micrometer/pull/3047#issuecomment-1137897921), I think the first option is more flexible.

As a reference, this is the difference in thresholds between 50% and 10%:
**50%**
```
  4MiB: 4_000
256MiB: 256_000
512MiB: 512_000
  1GiB: 1_024_000
 2+GiB: 2_000_000
```

**10%**
```
  4MiB: 0 // I will define a minimum for this, let's say 1_000
256MiB: 50_000
512MiB: 102_000
  1GiB: 204_000
  2GiB: 408_000
  4GiB: 818_000
  8GiB: 1_638_000
16+GiB: 2_000_000
```

Reply:
Sounds good to me.